infrastructure/src/main/java/org/corfudb/infrastructure/LayoutServer.java

```diff
@@ -232,56 +327,38 @@ public synchronized void handleMessageLayoutPropose(
-            r.sendResponse(ctx, msg, new CorfuPayloadMsg<>(CorfuMsgType.WRONG_EPOCH, serverEpoch));
-            log.trace("handleMessageLayoutPropose: Incoming message with wrong epoch, got {}, "
-                            + "expected {}, message was: {}",
-                    msg.getPayload().getEpoch(), serverEpoch, msg);
+        // Check if the prepare is for the correct epoch.
+        if (!isSameEpoch(msg.getPayload().getEpoch(), msg, ctx, r)) {
             return;
         }
+
+        final Rank messageRank = new Rank(msg.getPayload().getRank(), msg.getClientID());
+        final Rank highestProposedRank = getProposedRank();
+        final Layout messageValue = msg.getPayload().getLayout();
+
         // This is a propose. If no prepare, reject.
-        if (phase1Rank == null) {
-            log.debug("handleMessageLayoutPropose: Rejected phase 2 propose of rank={}, "
-                    + "phase1Rank=none", proposeRank);
+        if (highestProposedRank == null) {
+            log.debug("Phase II: Rejected propose of rank = {}, proposedRank = none",
+                    highestProposedRank);
             r.sendResponse(ctx, msg, CorfuMsgType.LAYOUT_PROPOSE_REJECT.payloadMsg(new
                     LayoutProposeResponse(-1)));
-            return;
-        }
-        // This is a propose. If the rank in the proposal is less than or equal to the highest yet
-        // observed prepare rank, reject.
-        if (!proposeRank.equals(phase1Rank)) {
-            log.debug("handleMessageLayoutPropose: Rejected phase 2 propose of rank={}, "
-                    + "phase1Rank={}", proposeRank, phase1Rank);
+        } else if (!messageRank.equals(highestProposedRank)) {
+            log.debug("Phase II: Rejected propose of rank = {}, proposedRank = {}",
+                    messageRank, highestProposedRank);
             r.sendResponse(ctx, msg, CorfuMsgType.LAYOUT_PROPOSE_REJECT.payloadMsg(new
-                    LayoutProposeResponse(phase1Rank.getRank())));
-            return;
-        }
-        Rank phase2Rank = getPhase2Rank();
-        Layout proposeLayout = msg.getPayload().getLayout();
-        // In addition, if the rank in the propose message is equal to the current phase 2 rank
-        // (already accepted message), reject.
-        // This can happen in case of duplicate messages.
-        if (proposeRank.equals(phase2Rank)) {
-            log.debug("handleMessageLayoutPropose: Rejected phase 2 propose of rank={}, "
-                    + "phase2Rank={}", proposeRank, phase2Rank);
-            r.sendResponse(ctx, msg, CorfuMsgType.LAYOUT_PROPOSE_REJECT.payloadMsg(new
-                    LayoutProposeResponse(phase2Rank.getRank())));
-            return;
+                    LayoutProposeResponse(highestProposedRank.getRank())));
+        } else {
+            log.debug("Phase II: New acceptedRank = {}, layout = {}",
+                    messageRank, messageValue);
+            setProposeRank(messageRank);
```

Comment:
Why you save phase1 rank one more time in spite of it was already saved before

Reply:
Yea, its not needed in this variant. Will remove it.